jenkins/python-setup.sh

```diff
@@ -17,10 +17,22 @@
 # chosen over any installed ones, and it adds `/usr/sbin` to the end of the
 # path for the `ip` command, used by pbench-register-tool.
 
-PATH=$(python3 -m site --user-base)/bin:${PATH}:/usr/sbin
+function _pbench_pathins() {
+    if [[ ":${PATH:-}:" != *":${1}:"* ]]; then
+        PATH="${1}${PATH:+:${PATH}}"
+    fi
+}
+
+_pbench_pathins "/usr/sbin"
+_pbench_pathins "/usr/bin"
+_pbench_pathins "$(python3 -m site --user-base)/bin"
+unset -f _pbench_pathins
+export PATH
+
 unset PYTHONPATH
-pip3 install --user -r lint-requirements.txt -r docs/requirements.txt -r agent/requirements.txt -r server/requirements.txt -r agent/test-requirements.txt -r server/test-requirements.txt
+pip3 install --user --no-cache-dir -r lint-requirements.txt -r docs/requirements.txt -r agent/requirements.txt -r server/requirements.txt -r agent/test-requirements.txt -r server/test-requirements.txt
 python3 setup.py develop --user
+export PYTHONPATH=$(python3 -c "import sys; print(':'.join(x for x in sys.path if x))")
```

Comment:
What is exporting `PYTHONPATH` supposed to accomplish here?

From `python --help`:
> PYTHONPATH   : ':'-separated list of directories prefixed to the default module search path.  The result is sys.path.

So, I think what this would do is double all of the non-blank entries in `sys.path`, except that it looks like Python suppresses the duplication, such that this has no effect at all.

Reply:
For any Python script run _after_ line 35 is executed, the value of `PYTHONPATH` is added to the front of the list of directories in `sys.path`; so, if we set `PYTHONPATH` to the contents of `sys.path`, then that would produce a list of those directories twice...except that, apparently, Python is clever enough to de-duplicate the list somehow.

As an experiment, I executed line 35 over and over, and always got the same result, which seemed weird, since I would have expected each subsequent execution to include the previous execution's result in the value of `sys.path`...but, each time, I just got the normal list (which, by the way, includes `~/.local`).

Reply:
Without this defined we get:

```
py run-test: commands[0] | black --check .
Traceback (most recent call last):
  File "/home/developer/.local/bin/black", line 5, in <module>
    from black import patched_main
ModuleNotFoundError: No module named 'black'
```

Reply:
Is this still true now that the virtual environment stuff is working?

Reply:
The container job does not use virtualenvs, `--current-env` turns that off so that all the pre-installed RPMs of the container are used.  Basically, the container IS the virtualenv.

Reply:
> The container job does not use virtualenvs

So that's what is stuck in my craw, here:  if we're running natively and not in a virtualenv, then setting the `PYTHONPATH` to `sys.path` should have no net effect!

If setting it like this is actually necessary in order to get something like Black to work, I'd like to understand why!...

Reply:
Certainly, but can we not have this hold up this work?

Reply:
I don't know how to answer, that, Peter.

Based on your experiments, we've found it necessary to add code which executes a statement which AFAICT should have no effect.  That sounds dangerously close to voodoo to me.

At the very least, we should have a nice fat comment explaining either why we are doing that or pointing out the fact that we need it and don't know why.  Otherwise, someone is going to come along later and spend a lot of time trying to delete it (or, at least, wondering why we're doing it).

Reply:
So, I've purchased our first clue.

Without an explicit `PYTHONPATH`, Black works fine in the container just before we invoke Tox, but it doesn't work inside the Tox invocation.  The difference is, outside of Tox, `sys.path` contains 
```
/usr/lib64/python39.zip:/usr/lib64/python3.9:/usr/lib64/python3.9/lib-dynload:/home/root/.local/lib/python3.9/site-packages:/usr/lib64/python3.9/site-packages:/usr/lib/python3.9/site-packages
```
but inside of Tox it contains
```
/usr/lib64/python39.zip:/usr/lib64/python3.9:/usr/lib64/python3.9/lib-dynload:/usr/lib64/python3.9/site-packages:/usr/lib/python3.9/site-packages
```
So, it's missing `/home/root/.local/lib/python3.9/site-packages` which is where the guts of Black have been installed.

Thus, by capturing the native `sys.path` in `PYTHONPATH`, we allow it to be inherited and used inside Tox, and that allows execution inside Tox to access stuff which wasn't installed there....

So, I think our Tox setup is missing a step.

Reply:
Hmm. Inside the container our local Python artifacts would be in the transient `/home/${USER}/.local`, and the in-container `python3 -m site` ought to see and report that path. This seems to be saying that's not happening; but `black` is actually there? That's weird... and I wonder if that's connected to the discussion about mucking with the container `${USER}` value. ☹️

Reply:
The sequence is:
- we enter the container
- we run, among other things, `python setup.py` which creates the `~/.local` installation (and Black works in this environment)
- we run `tox --current-env -e py`...and now things get fuzzier.  We're supposed to use the "current environment", but that no longer seems to include the `~/.local` installation.

We inherit the `PATH` value, which includes the Black executable/wrapper, but it cannot find its guts, because `sys.path` no longer includes `~/.local`.

`USER` is defined, and it has a reasonable value (`root`).

The problem is something to do with the Tox setup.  (I'm assuming that either we told it to do the wrong thing, or we didn't tell it everything to do.)  I think that defining `PYTHONPATH` outside and letting it be inherited is papering over the misconfiguration.

Reply:
The mystery has been unveiled.

Python adds the `--user` install directory to the `sys.path` any time that it exists (and/or is accessible).  However, it appears to locate the directory via the `HOME` environment variable.  And, it turns out that our `tox.ini` configuration was _not_ making that variable available to the test environment, so Python was not using there.  Once I fixed that, I was able to run Black (and Flake and PyTest) with this definition of `PYTHONPATH` commented out.

So, we can now remove this line from here, once we tweak `tox.ini`.